src/services/obs/backend.rs

```diff
@@ -0,0 +1,310 @@
+
+use anyhow::anyhow;
+use async_trait::async_trait;
+use http::StatusCode;
+use log::debug;
+use log::info;
+use reqsign::services::huaweicloud::obs::Signer;
+
+use crate::error::{other, BackendError};
+use crate::http_util::{
+    new_request_build_error, new_request_send_error, new_request_sign_error, parse_error_response,
+    percent_encode_path,
+};
+use crate::ops::BytesRange;
+use crate::services::util;
+use crate::Scheme;
+use crate::{
+    http_util::HttpClient,
+    ops::{OpCreate, OpDelete, OpList, OpRead, OpStat, OpWrite},
+    Accessor, AccessorMetadata, BytesReader, BytesWriter, DirStreamer, ObjectMetadata,
+};
+
+/// Builder for Huaweicloud OBS services
+#[derive(Default, Clone)]
+pub struct Builder {
+    root: Option<String>,
+    endpoint: Option<String>,
+    access_key: Option<String>,
+    secret_key: Option<String>,
+    bucket: Option<String>,
+    auth_bucket: Option<String>,
+}
+
+impl Debug for Builder {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        f.debug_struct("Builder")
+            .field("root", &self.root)
+            .field("endpoint", &self.endpoint)
+            .field("access_key", &"<redacted>")
+            .field("secret_key", &"<redacted>")
+            .field("bucket", &self.bucket)
+            .field("auth_bucket", &self.auth_bucket)
+            .finish()
+    }
+}
+
+impl Builder {
+    /// Set root of this backend.
+    ///
+    /// All operations will happen under this root.
+    pub fn root(&mut self, root: &str) -> &mut Self {
+        if !root.is_empty() {
+            self.root = Some(root.to_string())
+        }
+
+        self
+    }
+
+    /// Set endpoint of this backend.
+    ///
+    /// Endpoint must be full uri, e.g.
+    ///
+    /// - `https://bucketname.obs.cn-north-4.myhuaweicloud.com`
+    pub fn endpoint(&mut self, endpoint: &str) -> &mut Self {
+        if !endpoint.is_empty() {
+            self.endpoint = Some(endpoint.trim_end_matches('/').to_string());
+        }
+
+        self
+    }
+
+    /// Set access_key of this backend.
+    /// - If it is set, we will take user's input first.
+    /// - If not, we will try to load it from environment.
+    pub fn access_key(&mut self, access_key: &str) -> &mut Self {
+        if !access_key.is_empty() {
+            self.access_key = Some(access_key.to_string());
+        }
+
+        self
+    }
+
+    /// Set secret_key of this backend.
+    /// - If it is set, we will take user's input first.
+    /// - If not, we will try to load it from environment.
+    pub fn secret_key(&mut self, secret_key: &str) -> &mut Self {
+        if !secret_key.is_empty() {
+            self.secret_key = Some(secret_key.to_string());
+        }
+
+        self
+    }
+
+    /// Set bucket of this backend.
+    /// - If it is set, we will take user's input first.
+    /// - If not, we will try to load it from environment.
+    pub fn bucket(&mut self, bucket: &str) -> &mut Self {
```

Comment:
The `bucket` is required. Please raise an error if user doesn't input it.

Reply:
Currently we raise an error in Backend::build(), is it OK?

Reply:
LGTM.